keps/sig-network/2943-networkpolicy-status/README.md

```diff
@@ -0,0 +1,458 @@
+Policies not working in the cluster raise after she upgraded the cluster Network Policy provider to the latest 
+version.
+
+Users decided to create Network Policies with some fancy new features that were announced 
+in the release notes of a random NetPol provider, not knowing that this provider used in Samantha's cluster 
+is different and still cannot implement the new fields.
+
+Samantha needs to know exactly what are the complaints and why the provider cannot parse the 
+new rules, so she can explain to the users and also plan their monthly cluster upgrade to 
+add the required features.
+
+
+### Notes/Constraints/Caveats (Optional)
+The implementation of a status field depends on the CNI to provide this feedback. It's 
+not up to Kubernetes to know what happened, but to the controller of that object.
+
+If decided to use a "generic" approach of "controller / status / description" we cannot 
+guarantee that the CNI providers will follow the same standard/defaulting of the messages, 
+and this might be frustrating for users.
+
+### Risks and Mitigations
+Wrong RBAC in .status subresource field may lead to someone adding wrong status to users.
+
+The lack of the Network Policy Provider to provide a status to users may 
+lead to some confusion on whether the network policy could be properly parsed.
+
+It's up to the Network Policy Provider to decide how this feature is going to be 
+implemented, and this can be even an early admission webhook, making this feature
+useless.
+
+## Design Details
+API changes to NetworkPolicy:
+* Add the `status` inside `NetworkPolicy` struct as following:
+```go
+// NetworkPolicy describes what network traffic is allowed for a set of Pods
+type NetworkPolicy struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Specification of the desired behavior for this NetworkPolicy.
+	// +optional
+	Spec NetworkPolicySpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+
+  // Status is the current state of the NetworkPolicy.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
+	// +optional
+	Status NetworkPolicyStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+```
+
+* Add the `NetworkPolicyStatus` struct as following:
+```go
+// NetworkPolicyStatus describe the current state of the NetworkPolicy.
+type NetworkPolicyStatus struct {
+    // Conditions holds an array of metav1,Conditions that describe the state of the NetworkPolicy.
+  // When a CNI or other NetworkPolicy controller implementation parses this object,
+  // it must ensure that the `Accepted` Condition is updated with the parsing outcome.
```

Comment:
This mentions `Accepted` by name, but is that really what we want?  Or is that ALL we want?

As mentioned above, a single well-known Condition isn't really friendly to multiple simultaneous implementations (which is possible in this case, e.g. enforce in both iptables AND cloud firewall).  We could spec a dance of "Only set Accepted to `true` if `Accpeted` did not exist OR if the Reason and Message are yours (as defined by...?)".  That seems brittle but I don't know if it is worth adding multiple sets of Conditions  e.g.

```
Conditions []ProviderConditions
```

and 

```
type ProviderConditions struct {
  Provider string
  Conditions []metav1.Condition
}
```

One alternative could be to use the `Type` field as it is specced.  Instead of `Accepted` make it `myprovider.com/Accepted`.

But I am also not sure `Accepted` is the right polarity.  Why not `myprovider.com/PartialFailure` with `Reason` set to "FeaturesNotSupported" and `Message` like "Not supported: endPort".

@deads2k if he has "free time" to consult.  I think we can proceed with this KEP and argue over finer points of API at PR time.

Reply:
is network policy multiple provider a realistic problem?